source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,693 @@
+                        const std::vector<analysis::Const*>& operands) {
+  std::vector<uint32_t> operand_values_in_raw_words;
+  for (analysis::Const* operand : operands) {
+    if (analysis::ScalarConst* scalar = operand->AsScalarConst()) {
+      const auto& scalar_words = scalar->words();
+      assert(scalar_words.size() == 1 &&
+             "Scalar constants with longer than 32-bit width are not allowed "
+             "in OperateScalars()");
+      operand_values_in_raw_words.push_back(scalar_words.front());
+    } else if (operand->AsNullConst()) {
+      operand_values_in_raw_words.push_back(0u);
+    } else {
+      assert(false &&
+             "OperateScalars() only accepts ScalarConst or NullConst type of "
+             "constant");
+    }
+  }
+  return OperateWords(opcode, operand_values_in_raw_words);
+}
+
+// A wrapper function of the OperateWords() for constants of vector type. This
+// function iterates through the given vector type constant operands and
+// calculates the result for each element of the result vector to return.
+// Vectors with longer than 32-bit scalar components are not accepted in this
+// function.
+std::vector<uint32_t> OperateVectors(
+    SpvOp opcode, uint32_t num_dims,
+    const std::vector<analysis::Const*>& operands) {
+  std::vector<uint32_t> result;
+  for (uint32_t d = 0; d < num_dims; d++) {
+    std::vector<uint32_t> operand_values_for_one_dimension;
+    for (analysis::Const* operand : operands) {
+      if (analysis::VectorConst* vector_operand = operand->AsVectorConst()) {
+        if (const analysis::ScalarConst* scalar_component =
+                vector_operand->GetComponent(d)->AsScalarConst()) {
+          const auto& scalar_words = scalar_component->words();
+          assert(
+              scalar_words.size() == 1 &&
+              "Vector components with longer than 32-bit width are not allowed "
+              "in OperateVectors()");
+          operand_values_for_one_dimension.push_back(scalar_words.front());
+        } else if (operand->AsNullConst()) {
+          operand_values_for_one_dimension.push_back(0u);
+        } else {
+          assert(false &&
+                 "VectorConst should only has ScalarConst or NullConst as "
+                 "components");
+        }
+      } else if (operand->AsNullConst()) {
+        operand_values_for_one_dimension.push_back(0u);
+      } else {
+        assert(false &&
+               "OperateVectors() only accepts VectorConst or NullConst type of "
+               "constant");
+      }
+    }
+    result.push_back(OperateWords(opcode, operand_values_for_one_dimension));
+  }
+  return result;
+}
+}  // anonymous namespace
+
+bool FoldSpecConstantOpAndCompositePass::ProcessImpl(ir::Module* module) {
+  bool modified = false;
+  auto inst_iter = module->types_values_begin();
+  // In case the types, constants, global variabls secion is empty, do nothing.
+  if (inst_iter == module->types_values_end()) return modified;
+
+  do {
```

Comment:
Probably worth saying that you always need to re-evaluate the end iterator since you are modifying the list of instructions in this section of the module as you go along.